### ISSUE TITLE
Memory leak in core/adt

### ISSUE THREAD

**Issue by cueckoo** [cue-lang/cue#661]
_Originally opened by @samalba in https://github.com/cuelang/cue/issues/661_

### What version of CUE are you using (`cue version`)?

<pre>
$ cue version
0.3.0-beta.2
</pre>

### Does this issue reproduce with the latest release?

Yes. I also reproduced it as well with master (c58e0ffc4c74d1b3cffe6467b7f144b9dde631b7) - same result than beta.2.

### What did you do?

Cannot give a repro as it's happening with blocklayer's runtime, using the Cue Go API. In few seconds, the binary is consuming several gigs of RAM (stays stable at around 34MB with 0.3.0-alpha6).

As @myitcv suggested, I enabled pprof and gathered this attached heap profile.

[heap.out.gz](https://github.com/cuelang/cue/files/5858988/heap.out.gz)

Here are the top consumers:

```
(pprof) top
Showing nodes accounting for 1023.36MB, 87.95%!o(MISSING)f 1163.57MB total
Dropped 82 nodes (cum <= 5.82MB)
Showing top 10 nodes out of 84
      flat  flat%!s(MISSING)um%!c(MISSING)um   cum%!
(MISSING)
  322.54MB 27.72%   322.54MB 27.72%!c(MISSING)uelang.org/go/internal/core/adt.(*Vertex).GetArc
  158.51MB 13.62%   158.51MB 13.62%!c(MISSING)uelang.org/go/internal/core/adt.(*Vertex).addConjunct
  136.02MB 11.69%   136.02MB 11.69%!c(MISSING)uelang.org/go/internal/core/adt.(*nodeContext).createDisjunct
  113.01MB  9.71%   666.57MB 57.29%!c(MISSING)uelang.org/go/internal/core/adt.(*nodeContext).addStruct
   91.51MB  7.86%    91.51MB  7.86%!c(MISSING)uelang.org/go/internal/core/adt.updateCyclic
   75.50MB  6.49%    75.50MB  6.49%!c(MISSING)uelang.org/go/internal/core/adt.(*Vertex).AddStruct
   44.76MB  3.85%    50.27MB  4.32%!c(MISSING)uelang.org/go/pkg/encoding/json.Marshal
      39MB  3.35%    78.51MB  6.75%!c(MISSING)uelang.org/go/internal/core/adt.(*StructInfo).MatchAndInsert
      24MB  2.06%       24MB  2.06%!c(MISSING)uelang.org/go/internal/core/adt.clone
   18.50MB  1.59%    18.50MB  1.59%!c(MISSING)uelang.org/go/internal/core/adt.CloseInfo.SpawnRef (inline)
```

**Comment by cueckoo** [comment]
_Original reply by @myitcv in https://github.com/cuelang/cue/issues/661#issuecomment-765872181_

Very possible that https://github.com/cuelang/cue/issues/651 is the culprit here. Which has a fix pending https://cue-review.googlesource.com/c/cue/+/8282. A release fixing this (and #650 which is also fairly significant) is imminent.

**Comment by cueckoo** [comment]
_Original reply by @mpvl in https://github.com/cuelang/cue/issues/661#issuecomment-766099838_

The fix for #651 is in. You might want to try master.

If not, and since you can't give a reproducer, it would be good to bisect to determine the commit at which the issues occurred.

**Comment by cueckoo** [comment]
_Original reply by @samalba in https://github.com/cuelang/cue/issues/661#issuecomment-766110924_

Tried master just now and it looks like the issue is still there. Here is an updated pprof:

```
(pprof) top
Showing nodes accounting for 1947.71MB, 86.20%!o(MISSING)f 2259.62MB total
Dropped 152 nodes (cum <= 11.30MB)
Showing top 10 nodes out of 88
      flat  flat%!s(MISSING)um%!c(MISSING)um   cum%!
(MISSING)
  605.08MB 26.78%   605.08MB 26.78%!c(MISSING)uelang.org/go/internal/core/adt.(*Vertex).GetArc
  297.52MB 13.17%   297.52MB 13.17%!c(MISSING)uelang.org/go/internal/core/adt.(*Vertex).addConjunct
  276.03MB 12.22%   276.03MB 12.22%!c(MISSING)uelang.org/go/internal/core/adt.(*nodeContext).createDisjunct
  203.02MB  8.98%  1266.06MB 56.03%!c(MISSING)uelang.org/go/internal/core/adt.(*nodeContext).addStruct
  168.51MB  7.46%   168.51MB  7.46%!c(MISSING)uelang.org/go/internal/core/adt.updateCyclic
  158.51MB  7.01%   158.51MB  7.01%!c(MISSING)uelang.org/go/internal/core/adt.(*Vertex).AddStruct
   89.03MB  3.94%   110.06MB  4.87%!c(MISSING)uelang.org/go/pkg/encoding/json.Marshal
   70.51MB  3.12%   176.51MB  7.81%!c(MISSING)uelang.org/go/internal/core/adt.(*StructInfo).MatchAndInsert
   40.50MB  1.79%    40.50MB  1.79%!c(MISSING)uelang.org/go/internal/core/adt.CloseInfo.SpawnRef (inline)
   39.01MB  1.73%    39.01MB  1.73%!c(MISSING)uelang.org/go/internal/core/adt.clone
   ```

I'll spend more time to find the faulty commit. It's definitely somewhere between alpha6 and beta.1.

**Comment by cueckoo** [comment]
_Original reply by @myitcv in https://github.com/cuelang/cue/issues/661#issuecomment-766137489_

@samalba can you confirm exactly what commit you tested?

**Comment by cueckoo** [comment]
_Original reply by @samalba in https://github.com/cuelang/cue/issues/661#issuecomment-766137917_

> @samalba can you confirm exactly what commit you tested?

Yes sorry, it is 0e401d66c8e5a7e64c667dfa64004cadca7cb871 (tagged v0.3.0-beta.3)

**Comment by cueckoo** [comment]
_Original reply by @samalba in https://github.com/cuelang/cue/issues/661#issuecomment-766166759_

I found the culprit: 008d37d24e452beadafbba5220cf757d66a3bbe3. I confirm the problem does not occur with previous commits.

**Comment by cueckoo** [comment]
_Original reply by @mpvl in https://github.com/cuelang/cue/issues/661#issuecomment-766171384_

Thanks for the bisect. That is very surprising, and also somewhat encouraging. That change is fairly small and also immaterial for correctness. It is mostly meant for better and less confusing error messages.

You could try to comment out the body of `addNotify`(3 lines) at `cuelang.org/go/internal/core/adt/eval.go:724` and see what that changes for you.

**Comment by cueckoo** [comment]
_Original reply by @mpvl in https://github.com/cuelang/cue/issues/661#issuecomment-766172966_

I did find a bug with that CL. It seems to improve performance a bit if I add
```
			notify:        n.notify[:0],
```
to `newNodeContext` in `eval.go` around `777`. I'm not sure if that explains the full extent of the slowdown you're seeing, but it is a possibility.

**Comment by cueckoo** [comment]
_Original reply by @samalba in https://github.com/cuelang/cue/issues/661#issuecomment-766180516_

> You could try to comment out the body of `addNotify`(3 lines) at `cuelang.org/go/internal/core/adt/eval.go:724` and see what that changes for you.

Yep, I confirm that the leak disappear if I comment out the content of the addNotify.

**Comment by cueckoo** [comment]
_Original reply by @mpvl in https://github.com/cuelang/cue/issues/661#issuecomment-766187106_

Great! Then I think just adding the second line will work. I'll take a close look and send out a CL. Thanks!

**Comment by cueckoo** [comment]
_Original reply by @mpvl in https://github.com/cuelang/cue/issues/661#issuecomment-766358491_

I've checked in what I think will fix it, while keeping the good error messages. Could you please verify? Thanks!

**Comment by cueckoo** [comment]
_Original reply by @samalba in https://github.com/cuelang/cue/issues/661#issuecomment-766452072_

I tested with the commit 9c5489fc66edf789372390847d4c3e10f047a005 and unfortunately, the leak is still present. I commented the addNotify() again and it fixes the leak.

**Comment by cueckoo** [comment]
_Original reply by @myitcv in https://github.com/cuelang/cue/issues/661#issuecomment-766693149_

@samalba let's catch up later to work out a plan for coming up with a reproducer.

**Comment by cueckoo** [comment]
_Original reply by @samalba in https://github.com/cuelang/cue/issues/661#issuecomment-774325354_

Giving an update where we are: we replaced our home-made task detection with cueflow which considerably reduced the our calls to the cue go API. As a result, the leak disappeared. And we were able to upgrade to 0.3.0-beta.3 in production without any issue. The problem is probably still there but it's no more a blocker for us, I cannot even reproduce it with the new code base.
